File: backend/app/api/routes/analytics.py
"""Analytics endpoints: one aggregated dashboard read plus activity feeds.

Read-only by design (no POST/PUT/DELETE here). Every route verifies
authenticated user -> owned project; cross-tenant access returns 404 without
disclosure. Window clamps keep every query bounded.
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.auth.dependencies import CurrentUser, get_owned_project
from app.db.session import get_db
from app.models.enums import EventType
from app.models.learning import Project
from app.schemas.analytics import (
    ActivityDayRead,
    ActivityItemRead,
    DashboardSummaryRead,
    EventRead,
    MasteryTrendPointRead,
)
from app.schemas.common import Page
from app.services.analytics_service import AnalyticsService
from app.services.event_service import EventService

router = APIRouter(prefix="/projects", tags=["analytics"])

SessionDep = Annotated[Session, Depends(get_db)]
OwnedProject = Annotated[Project, Depends(get_owned_project)]
RangeParam = Annotated[
    Literal["7d", "30d", "90d", "all"], Query(description="UTC window for activity data")
]


def _to_activity(item) -> ActivityItemRead:
    return ActivityItemRead(
        id=item.id,
        event_type=item.event_type,
        created_at=item.created_at,
        resource_id=item.resource_id,
        metadata=item.metadata,
        summary=item.summary,
    )


@router.get("/{project_id}/analytics/dashboard", response_model=DashboardSummaryRead)
def get_dashboard(project: OwnedProject, user: CurrentUser, session: SessionDep):
    service = AnalyticsService(session)
    summary = service.dashboard_summary(user_id=user.id, project_id=project.id)
    return DashboardSummaryRead(
        project_id=summary.project_id,
        project_name=summary.project_name,
        materials_count=summary.materials_count,
        materials_ready=summary.materials_ready,
        materials_failed=summary.materials_failed,
        documents_count=summary.documents_count,
        pages_count=summary.pages_count,
        chunks_count=summary.chunks_count,
        images_count=summary.images_count,
        concepts_count=summary.concepts_count,
        assessment_count=summary.assessment_count,
        questions_answered=summary.questions_answered,
        questions_correct=summary.questions_correct,
        questions_partial=summary.questions_partial,
        questions_incorrect=summary.questions_incorrect,
        average_assessment_score=summary.average_assessment_score,
        overall_mastery=summary.overall_mastery,
        mastery_confidence=summary.mastery_confidence,
        growth_status=summary.growth_status,
        active_recommendations=summary.active_recommendations,
        tutor_conversations=summary.tutor_conversations,
        tutor_messages=summary.tutor_messages,
        last_activity_at=summary.last_activity_at,
        has_learning_evidence=summary.has_learning_evidence,
    )


@router.get("/{project_id}/analytics/activity", response_model=Page[ActivityItemRead])
def get_activity(
    project: OwnedProject,
    user: CurrentUser,
    session: SessionDep,
    range: RangeParam = "30d",
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    event_type: Annotated[list[EventType] | None, Query()] = None,
):
    service = AnalyticsService(session)
    items = service.activity(
        user_id=user.id,
        project_id=project.id,
        event_types=event_type,
        window=range,
        limit=limit,
    )
    return Page[ActivityItemRead](
        items=[_to_activity(item) for item in items],
        total=len(items),
        page=1,
        page_size=limit,
    )


@router.get("/{project_id}/analytics/activity-by-day", response_model=list[ActivityDayRead])
def get_activity_by_day(
    project: OwnedProject, user: CurrentUser, session: SessionDep, range: RangeParam = "30d"
):
    service = AnalyticsService(session)
    days = service.activity_by_day(user_id=user.id, project_id=project.id, window=range)
    return [ActivityDayRead(date=day.date, count=day.count) for day in days]


@router.get("/{project_id}/analytics/mastery-trend", response_model=list[MasteryTrendPointRead])
def get_mastery_trend(
    project: OwnedProject,
    user: CurrentUser,
    session: SessionDep,
    limit: Annotated[int, Query(ge=1, le=100)] = 30,
):
    service = AnalyticsService(session)
    points = service.mastery_trend(user_id=user.id, project_id=project.id, limit=limit)
    return [
        MasteryTrendPointRead(date=p.date, score=p.score, assessment_id=p.assessment_id)
        for p in points
    ]


@router.get("/{project_id}/events", response_model=Page[EventRead])
def list_events(
    project: OwnedProject,
    user: CurrentUser,
    session: SessionDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
    event_type: Annotated[list[EventType] | None, Query()] = None,
    since: Annotated[datetime | None, Query(description="UTC lower bound (inclusive)")] = None,
    until: Annotated[datetime | None, Query(description="UTC upper bound (inclusive)")] = None,
):
    """Project activity feed (tenant-scoped, newest first, offset pages)."""
    service = EventService(session)
    rows, total = service.feed(
        user_id=user.id,
        project_id=project.id,
        event_types=event_type,
        since=since,
        until=until,
        page=page,
        page_size=page_size,
    )
    from app.services.event_service import sanitize_payload

    return Page[EventRead](
        items=[
            EventRead(
                id=row.id,
                event_type=row.event_type,
                created_at=row.created_at,
                project_id=row.project_id,
                resource_id=row.entity_id,
                metadata=sanitize_payload(row.payload),
            )
            for row in rows
        ],
        total=total,
        page=page,
        page_size=page_size,
    )

File: frontend/src/api/analytics.ts
import { apiClient } from "./client";

export type DateRange = "7d" | "30d" | "90d" | "all";

export interface DashboardSummary {
  projectId: string;
  projectName: string;
  materialsCount: number;
  materialsReady: number;
  materialsFailed: number;
  documentsCount: number;
  pagesCount: number;
  chunksCount: number;
  imagesCount: number;
  conceptsCount: number;
  assessmentCount: number;
  questionsAnswered: number;
  questionsCorrect: number;
  questionsPartial: number;
  questionsIncorrect: number;
  averageAssessmentScore: number | null;
  overallMastery: number | null;
  masteryConfidence: number | null;
  growthStatus: "IMPROVING" | "STABLE" | "REQUIRING_ATTENTION" | null;
  activeRecommendations: number;
  tutorConversations: number;
  tutorMessages: number;
  lastActivityAt: string | null;
  hasLearningEvidence: boolean;
}

export interface ActivityItem {
  id: string;
  eventType: string;
  createdAt: string;
  resourceId: string | null;
  metadata: Record<string, unknown> | null;
  summary: string;
}

export interface ActivityDay {
  date: string;
  count: number;
}

export interface MasteryTrendPoint {
  date: string;
  score: number;
  assessmentId: string;
}

interface BackendSummary {
  project_id: string;
  project_name: string;
  materials_count: number;
  materials_ready: number;
  materials_failed: number;
  documents_count: number;
  pages_count: number;
  chunks_count: number;
  images_count: number;
  concepts_count: number;
  assessment_count: number;
  questions_answered: number;
  questions_correct: number;
  questions_partial: number;
  questions_incorrect: number;
  average_assessment_score: number | null;
  overall_mastery: number | null;
  mastery_confidence: number | null;
  growth_status: DashboardSummary["growthStatus"];
  active_recommendations: number;
  tutor_conversations: number;
  tutor_messages: number;
  last_activity_at: string | null;
  has_learning_evidence: boolean;
}

function toSummary(s: BackendSummary): DashboardSummary {
  return {
    projectId: s.project_id,
    projectName: s.project_name,
    materialsCount: s.materials_count,
    materialsReady: s.materials_ready,
    materialsFailed: s.materials_failed,
    documentsCount: s.documents_count,
    pagesCount: s.pages_count,
    chunksCount: s.chunks_count,
    imagesCount: s.images_count,
    conceptsCount: s.concepts_count,
    assessmentCount: s.assessment_count,
    questionsAnswered: s.questions_answered,
    questionsCorrect: s.questions_correct,
    questionsPartial: s.questions_partial,
    questionsIncorrect: s.questions_incorrect,
    averageAssessmentScore: s.average_assessment_score,
    overallMastery: s.overall_mastery,
    masteryConfidence: s.mastery_confidence,
    growthStatus: s.growth_status,
    activeRecommendations: s.active_recommendations,
    tutorConversations: s.tutor_conversations,
    tutorMessages: s.tutor_messages,
    lastActivityAt: s.last_activity_at,
    hasLearningEvidence: s.has_learning_evidence,
  };
}

interface BackendActivityItem {
  id: string;
  event_type: string;
  created_at: string;
  resource_id: string | null;
  metadata: Record<string, unknown> | null;
  summary: string;
}

function toActivityItem(a: BackendActivityItem): ActivityItem {
  return {
    id: a.id,
    eventType: a.event_type,
    createdAt: a.created_at,
    resourceId: a.resource_id,
    metadata: a.metadata,
    summary: a.summary,
  };
}

export const analyticsApi = {
  async dashboard(projectId: string): Promise<DashboardSummary> {
    const res = await apiClient.get<BackendSummary>(
      `/api/v1/projects/${projectId}/analytics/dashboard`,
    );
    return toSummary(res.data);
  },

  async activity(
    projectId: string,
    params: { range?: DateRange; limit?: number } = {},
  ): Promise<ActivityItem[]> {
    const res = await apiClient.get<{ items: BackendActivityItem[] }>(
      `/api/v1/projects/${projectId}/analytics/activity`,
      { params: { range: params.range ?? "30d", limit: params.limit ?? 8 } },
    );
    return res.data.items.map(toActivityItem);
  },

  async activityByDay(projectId: string, range: DateRange = "30d"): Promise<ActivityDay[]> {
    const res = await apiClient.get<ActivityDay[]>(
      `/api/v1/projects/${projectId}/analytics/activity-by-day`,
      { params: { range } },
    );
    return res.data;
  },

  async masteryTrend(projectId: string): Promise<MasteryTrendPoint[]> {
    const res = await apiClient.get<{ date: string; score: number; assessment_id: string }[]>(
      `/api/v1/projects/${projectId}/analytics/mastery-trend`,
    );
    return res.data.map((p) => ({ date: p.date, score: p.score, assessmentId: p.assessment_id }));
  },
};
